decisions/0003-files-in-JSON-output.md

```diff
@@ -121,27 +144,183 @@ lost
+Builder is notified that the file has been successfully retrieved and stored
+elsewhere as knowing we've served the file successfully doesn't guarantee the
+other side has it.
+
+### Issues concerning the revoking if signed URLs:
+
+S3 signed URLs expire after a given time, and not after a number of uses - if we
+want to do the latter we would have to implement a proxy application
+
+Revoking a S3 signed URL on demand involves removing the permissions of the IAM
+user who signed the URL, so we couldn't expire individual URLs ourselves after
+we know they've been used successfully unless we also generate a new IAM user to
+sign each URL for each file. This is presumably possible but not ideal. So
+revoking signed URLs on demand may also mean using a proxy application.
+
+CMSs which retrieve files directly themselves are unlikely to be able to notify
+us when they've successfully fetched the file - a decrypting adapter could make
+this callback, but probably wouldn't know whether the CMS had stored the file.
+
+This whole question means that Form Builder needs not to care too much about
+what's happening a long way downstream.
+
+### Conclusion
+
+Considering all the points above option 6 has been chosen.
+
+It leaves a clear cut boundary between Form Builder and integrations. Although
+the responsibility of decrypting files is now in the hands of the receiver. The
+Form Builder team does not need a proxy application and S3 can be left to handle
+files.
+
+It deals with encryption: files are kept encrypted at rest and only decrypted
+when being processed or exiting the system.
+
+Signed URLs are less of a concern as the files themselves are encrypted as they
+leave S3.
+
+### First use case
+
+For HMCTS Complaints Form this would mean that the CMS would have to fetch the
+files through the adapter so they can be decrypted. So this is what a JSON
+submission including uploaded files looks like:
+
+- Form Builder makes JSON request (including S3 signed URLs) to adapter
+- The adapter decrypts the JSON, and POSTs the submitted data to the CMS,
+including adapter links to the files
+- CMS requests the files from the adapter
+- The adapter fetches the file from S3, decrypts it, and responds with it to the
+CMS
+- CMS stores the files
+
+This option takes a clear-cut approach to responsibilities and security. It's
+one of the more feasible options in the timescale we have for HMCTS complaints,
+since it doesn't involve building a proxy application.
 
 ## Consequences
 
-- Other systems can now easily integrate with form builder to further process
-user submissions including files uploaded
-- As signed URLs are being generated if this information were to be leaked,
-they would be able to access user files and lead back to form builder
+### Risks
+
+As for creating any API allowing for integrations there will always be the off
+chance that if any information is obtained by unintended audiences they could be
+able to access files.
```

Comment:
The risks of sending the JSON payload to the wrong place or it being intercepted are mitigated by encrypting all transferred data with the PSK.
The risk of requests to the receiver being forged is mitigated by encrypting all transferred data with the PSK.
The risk of unauthorised people attempting to traverse/list the files in the bucket to access them is mitigated by using signed URLs and configuring the bucket well.
The risk of the receiver accidentally gaining access to the wrong files in the bucket is mitigated by using signed URLs and encrypting all transferred data with the PSK.

Reply:
Do you want me to add the above? or are you questioning the paragraph?

If the latter then you've made the assumption that an attacker was only able to obtain one piece of information. Therefore with our processes they have been thwarted. I'm saying at a high level that if someone was able to gain several pieces of information they could be able to read the information. This was only made possible by the fact we are implementing a feature that shares data to other parties which is inherently a risk.